pkg/controller/noderefutil/providerid_test.go

```diff
@@ -17,61 +17,109 @@ limitations under the License.
 package noderefutil
 
 import (
-	"strings"
 	"testing"
-
-	"github.com/pkg/errors"
 )
 
+const aws = "aws"
+
 func TestNewProviderID(t *testing.T) {
-	input1 := "aws:////instance-id"
-	_, err := NewProviderID(input1)
-	if err != nil {
-		t.Fatalf("Expected no errors, got %v", err)
+	tests := []struct {
+		name       string
+		input      string
+		expectedID string
+	}{
+		{
+			name:       "2 slashes after colon, one segment",
+			input:      "aws://instance-id",
+			expectedID: "instance-id",
+		},
+		{
+			name:       "more than 2 slashes after colon, one segment",
+			input:      "aws:////instance-id",
+			expectedID: "instance-id",
+		},
+		{
+			name:       "multiple filled-in segments (aws format)",
+			input:      "aws:///zone/instance-id",
+			expectedID: "instance-id",
+		},
+		{
+			name:       "multiple filled-in segments",
+			input:      "aws://bar/baz/instance-id",
+			expectedID: "instance-id",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			id, err := NewProviderID(tc.input)
+			if err != nil {
+				t.Fatalf("Expected no errors, got %v", err)
+			}
+
+			if id.CloudProvider() != aws {
+				t.Errorf("Unexpected cloud provider: %q", id.CloudProvider())
+			}
+
+			if e, a := tc.expectedID, id.ID(); e != a {
+				t.Errorf("Expected %q, got %q", e, a)
+			}
+		})
 	}
+
 }
 
 func TestInvalidProviderID(t *testing.T) {
 	testCases := []struct {
+		name  string
 		input string
 		err   error
 	}{
 		{
+			name:  "empty id",
+			input: "",
+			err:   ErrEmptyProviderID,
+		},
+		{
+			name:  "only empty segments",
 			input: "aws:///////",
 			err:   ErrInvalidProviderID,
 		},
 		{
-			input: ":///instance-id",
-			err:   errors.New("missing protocol scheme"),
+			name:  "missing cloud provider",
+			input: "://instance-id",
+			err:   ErrInvalidProviderID,
 		},
 		{
-			input: "///instance-id",
+			name:  "missing cloud provider and colon",
+			input: "//instance-id",
 			err:   ErrInvalidProviderID,
 		},
 		{
+			name:  "missing cloud provider, colon, one leading slash",
 			input: "/instance-id",
 			err:   ErrInvalidProviderID,
 		},
 		{
+			name:  "just an id",
 			input: "instance-id",
 			err:   ErrInvalidProviderID,
 		},
-		{
-			input: "#IAmTheSenate",
-			err:   ErrInvalidProviderID,
-		},
 	}
 
 	for _, test := range testCases {
-		_, err := NewProviderID(test.input)
-		if !strings.Contains(err.Error(), test.err.Error()) {
-			t.Fatalf("Expected error %v, got %v", test.err, err)
-		}
+		t.Run(test.name, func(t *testing.T) {
+			_, err := NewProviderID(test.input)
+			if test.err != err {
+				t.Fatalf("Expected error %v, got %v", test.err, err)
+			}
+		})
+
 	}
 }
 
 func TestProviderIDEquals(t *testing.T) {
-	input1 := "aws:////instance-id1"
+	input1 := "aws://instance-id1"
```

Comment:
could tablify this test, i think having multiple equals examples would be helpful here

could also add a separate TestFunction for testing inequality, but that is extra credit and super not necessary